_posts/2019-05-31-ucx-dgx.md

```diff
@@ -0,0 +1,531 @@
+from distributed.worker import TOTAL_MEMORY
+
+from dask_cuda.local_cuda_cluster import cuda_visible_devices
+
+
+class CPUAffinity:
+    """ A Worker plugin to pin CPU affinity """
+    def __init__(self, cores):
+        self.cores = cores
+
+    def setup(self, worker=None):
+        os.sched_setaffinity(0, self.cores)
+
+
+affinity = {  # See nvidia-smi topo -m
+    0: list(range(0, 20)) + list(range(40, 60)),
+    1: list(range(0, 20)) + list(range(40, 60)),
+    2: list(range(0, 20)) + list(range(40, 60)),
+    3: list(range(0, 20)) + list(range(40, 60)),
+    4: list(range(20, 40)) + list(range(60, 79)),
+    5: list(range(20, 40)) + list(range(60, 79)),
+    6: list(range(20, 40)) + list(range(60, 79)),
+    7: list(range(20, 40)) + list(range(60, 79)),
+}
+
+def DGX(
+    interface="ib",
+    dashboard_address=":8787",
+    threads_per_worker=1,
+    silence_logs=True,
+    CUDA_VISIBLE_DEVICES=None,
+    **kwargs
+):
+    """ A Local Cluster for a DGX 1 machine
+
+    NVIDIA's DGX-1 machine has a complex architecture mapping CPUs, GPUs, and
+    network hardware.  This function creates a local cluster that tries to
+    respect this hardware as much as possible.
+
+    It creates one Dask worker process per GPU, and assigns each worker process
+    the correct CPU cores and Network interface cards to maximize performance.
+
+    That being said, things aren't perfect.  Today a DGX has very high
+    performance between certain sets of GPUs and not others.  A Dask DGX
+    cluster that uses only certain tightly coupled parts of the computer will
+    have significantly higher bandwidth than a deployment on the entire thing.
+
+    Parameters
+    ----------
+    interface: str
+        The interface prefix for the infiniband networking cards.  This is
+        often "ib"` or "bond".  We will add the numeric suffix 0,1,2,3 as
+        appropriate.  Defaults to "ib".
+    dashboard_address: str
+        The address for the scheduler dashboard.  Defaults to ":8787".
+    CUDA_VISIBLE_DEVICES: str
+        String like ``"0,1,2,3"`` or ``[0, 1, 2, 3]`` to restrict activity to
+        different GPUs
+
+    Examples
+    --------
+    >>> from dask_cuda import DGX
+    >>> from dask.distributed import Client
+    >>> cluster = DGX(interface='ib')
+    >>> client = Client(cluster)
+    """
+    if CUDA_VISIBLE_DEVICES is None:
+        CUDA_VISIBLE_DEVICES = os.environ.get("CUDA_VISIBLE_DEVICES", "0,1,2,3,4,5,6,7")
+    if isinstance(CUDA_VISIBLE_DEVICES, str):
+        CUDA_VISIBLE_DEVICES = CUDA_VISIBLE_DEVICES.split(",")
+    CUDA_VISIBLE_DEVICES = list(map(int, CUDA_VISIBLE_DEVICES))
+    memory_limit = TOTAL_MEMORY / 8
+
+    spec = {
+        i: {
+            "cls": Nanny,
+            "options": {
+                "env": {
+                    "CUDA_VISIBLE_DEVICES": cuda_visible_devices(
+                        ii, CUDA_VISIBLE_DEVICES
+                    ),
+                    # 'UCX_NET_DEVICES': 'mlx5_%d:1' % (i // 2)
```

Comment:
Should we keep this comment? If so, could we explain what it is here for with another comment?